fs/nfs/nfs4proc.c

```diff
@@ -7771,6 +7778,488 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
+ *
+ * Upon success 0 is returned, and the data parameter is advanced.
+ */
+static int nfs4_decode_nfsace4(void **data, struct nfs4_ace *ace)
+{
+	u32 *p = *data;
+	u32 length;
+	int error = 0;
+
+	/*
+	 * Per RFC8881 6.2.1 acl the format of nfsace4 is:
+	 * uint32_t type
+	 * uint32_t flag
+	 * uint32_t access_mask
+	 * utf8str_mixed who
+	 *
+	 * where utf8str_mixed is a utf8string.  Per RFC 1832
+	 * this is encoded as a uint32_t length followed by
+	 * string bytes, and then padded by null up to a 4 byte
+	 * boundary.
+	 */
+	ace->type = ntohl(*(p++));
+	ace->flag = ntohl(*(p++));
+	ace->access_mask = ntohl(*(p++));
+	length = ntohl(*(p++));
+
+	ace->whotype = nfs4_acl_get_whotype((char *)p, length);
+
+	if (ace->whotype == NFS4_ACL_WHO_NAMED) {
+		/*
+		 * We're just going to support numeric IDs here.
+		 * (it's very expensive to do otherwise)
+		 */
+		unsigned long id = -1;
+
+		if (numeric_name_to_id((const char *)p, length, &id)) {
+			if (ace->flag & NFS4_ACE_IDENTIFIER_GROUP) {
+				ace->who_gid = make_kgid(&init_user_ns, id);
+				if (!gid_valid(ace->who_gid)) {
+					pr_err("NFS: %s Invalid who_gid: %.*s", __func__,
+					       length, (char *)p);
+					error = -EINVAL;
+				}
+			} else {
+				ace->who_uid = make_kuid(&init_user_ns, id);
+				if (!uid_valid(ace->who_uid)) {
+					pr_err("NFS: %s Invalid who_uid: %.*s", __func__,
+					       length, (char *)p);
+					error = -EINVAL;
+				}
+			}
+		} else {
+			pr_err("NFS: %s Non-numeric who: %.*s", __func__, length, (char *)p);
+			error = -EINVAL;
+		}
+	}
+	/*
+	 * If the whotype was a value other than NFS4_ACL_WHO_NAMED
+	 * (e.g. NFS4_ACL_WHO_OWNER), then no further decoding is
+	 * required (so, no else statement).
+	 */
+
+	if (error >= 0) {
+		/*
+		 * Advance by a multiple of 4 bytes (string + padding).
+		 * Since p is a pointer to u32 we don't need to << 2 the
+		 * output from XDR_QUADLEN.
+		 */
+		p += XDR_QUADLEN(length);
+		*data = p;
+	}
+	return error;
+}
+
+/*
+ * Decode the buffer containing the raw XDR data into struct nfs4_acl.
+ *
+ * This is similar to nfsd4_decode_acl
+ */
+static int nfs4_aclbuf_decode(void *buf, size_t buflen, enum nfs4_acl_type type,
+			      struct nfs4_acl **pacl)
+{
+	u32 acl_flag, count;
+	u32 *p = buf;
+	struct nfs4_acl *acl;
+	int acl_bytes;
+	struct nfs4_ace *ace;
+	int status;
+
+	if (buflen < 8) {
+		pr_err("NFS: %s ACL buffer too small: %zd", __func__, buflen);
+		return -EINVAL;
+	}
+
+	acl_flag = ntohl(*(p++));
+	count = ntohl(*(p++));
+	acl_bytes = nfs4_acl_bytes(count);
+	acl = kmalloc(acl_bytes, GFP_KERNEL);
+	if (acl == NULL)
+		return -ENOMEM;
+
+	acl->flag = acl_flag;
+	acl->naces = count;
+	for (ace = acl->aces; ace < acl->aces + count; ace++) {
+		/*
+		 * Ensure that we have not overrun our input buffer
+		 * (Output buffer controlled by for loop condition)
+		 */
+		if ((void *)p >= (buf + buflen))
+			status = -EINVAL;
+		else
+			status = nfs4_decode_nfsace4((void **)&p, ace);
```

Comment:
nfs4_decode_nfsace4() increments p by multiple words, so the check in its present form is insufficient.  I wonder if it could be cleaner if we pass remaining buffer size to nfs4_decode_nfsace4() to check it inside.